### PR TITLE
enh(autologin): verify that autologin is different than password

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR intends to verify that autologin is different than password
**Fixes** # MON-12050

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/include/Administration/myAccount/DB-Func.php [hunk 1]
```diff
@@ -219,3 +219,38 @@ function validatePasswordModification(array $fields)
...
+        $statement = $pearDB->prepare(
+            'SELECT * FROM `contact_password` WHERE contact_id = :contactId ORDER BY creation_date DESC LIMIT 1'
+        );
+        $statement->bindValue(':contactId', $contactId, \PDO::PARAM_INT);
+        $statement->execute();
+
+        if (
+            ($result = $statement->fetch(\PDO::FETCH_ASSOC))
+            && password_verify($fields['contact_autologin_key'], $result['password'])
+            && !empty($fields['contact_passwd'])
```
Comment (kduret) [inline www/include/Administration/myAccount/DB-Func.php:242]:
```suggestion
            ($result = $statement->fetch(\PDO::FETCH_ASSOC))
            && !empty($fields['contact_passwd'])
            && password_verify($fields['contact_autologin_key'], $result['password'])
```

File: lib/Centreon/Object/Contact/Contact.php [hunk 2]
```diff
@@ -163,6 +178,20 @@ public function update($contactId, $params = array())
...
+                "SELECT password FROM contact_password WHERE contact_id = :contactId " .
+                "ORDER BY creation_date DESC LIMIT 1"
+            );
+            $statement->bindValue(':contactId', $contactId, \PDO::PARAM_INT);
+            $statement->execute();
+            if (
+                ($result = $statement->fetch(\PDO::FETCH_ASSOC))
+                && password_verify($params['contact_autologin_key'], $result['password'])
+            ) {
+                throw new \Exception('Your autologin key should be different than your current password');
```
Comment (kduret) [inline lib/Centreon/Object/Contact/Contact.php:192]:
```suggestion
                throw new \Exception(_('Your autologin key should be different than your current password'));
```

File: www/include/Administration/myAccount/DB-Func.php [hunk 3]
```diff
@@ -219,3 +219,38 @@ function validatePasswordModification(array $fields)
...
+        ) {
+            $errors['contact_autologin_key'] = _('Your autologin key should be different than your current password');
+        } elseif (
+            empty($fields['contact_passwd'])
+            && $fields['contact_passwd'] === $fields['contact_autologin_key']
+        ) {
+            $errorMessage = 'Your new password and autologin key should be different';
+            $errors['contact_passwd'] = _($errorMessage);
+            $errors['contact_autologin_key'] = _($errorMessage);
```
Comment (kduret) [inline www/include/Administration/myAccount/DB-Func.php:251]:
```suggestion
            $errorMessage = _('Your new password and autologin key should be different');
            $errors['contact_passwd'] = $errorMessage;
            $errors['contact_autologin_key'] = $errorMessage;
```

File: www/include/configuration/configObject/contact/DB-Func.php [hunk 4]
```diff
@@ -1322,3 +1322,48 @@ function validatePasswordModification(array $fields)
 
     return count($errors) > 0 ? $errors : true;
 }
+
+/**
+ * Validate autologin key is not equal to a password
+ *
+ * @param array $fields
+ * @return mixed
```
Comment (kduret) [inline www/include/configuration/configObject/contact/DB-Func.php:1330]:
```suggestion
 * @return array<string,string>|bool
```

File: www/include/Administration/myAccount/DB-Func.php [hunk 5]
```diff
@@ -219,3 +219,38 @@ function validatePasswordModification(array $fields)
 
     return count($errors) > 0 ? $errors : true;
 }
+
+/**
+ * @param array $fields
```
Comment (kduret) [inline www/include/Administration/myAccount/DB-Func.php:224]:
```suggestion
 * @param array $fields
 * @return array<string,string>|bool
```

File: www/include/Administration/myAccount/DB-Func.php [hunk 6]
```diff
@@ -219,3 +219,38 @@ function validatePasswordModification(array $fields)
...
+        $statement->execute();
+
+        if (
+            ($result = $statement->fetch(\PDO::FETCH_ASSOC))
+            && password_verify($fields['contact_autologin_key'], $result['password'])
+            && !empty($fields['contact_passwd'])
+        ) {
+            $errors['contact_autologin_key'] = _('Your autologin key should be different than your current password');
+        } elseif (
+            empty($fields['contact_passwd'])
```
Comment (callapa) [inline www/include/Administration/myAccount/DB-Func.php:246]:
```suggestion
            !empty($fields['contact_passwd'])
```

**Review by callapa — CHANGES_REQUESTED** [review]

File: www/include/Administration/myAccount/DB-Func.php [hunk 7]
```diff
@@ -219,3 +219,39 @@ function validatePasswordModification(array $fields)
...
+            !empty($fields['contact_passwd'])
+            && $fields['contact_passwd'] === $fields['contact_autologin_key']
+        ) {
+            $errorMessage = _('Your new password and autologin key should be different');
+            $errors['contact_passwd'] = $errorMessage;
+            $errors['contact_autologin_key'] = $errorMessage;
+        }
+    }
+
+    return count($errors) > 0 ? $errors : true;
```
Comment (callapa) [inline www/include/Administration/myAccount/DB-Func.php:256]: You cannot return an array or a Boolean value, you must make a choice.

File: www/include/configuration/configObject/contact/DB-Func.php [hunk 8]
```diff
@@ -1322,3 +1322,48 @@ function validatePasswordModification(array $fields)
...
+            !empty($fields['contact_passwd'])
+            && $fields['contact_passwd'] === $fields['contact_autologin_key']
+        ) {
+            $errorMessage = 'Your password and autologin key should be different';
+            $errors['contact_passwd'] = _($errorMessage);
+            $errors['contact_autologin_key'] = _($errorMessage);
+        }
+    }
+
+    return count($errors) > 0 ? $errors : true;
```
Comment (callapa) [inline www/include/configuration/configObject/contact/DB-Func.php:1368]: You cannot return an array or a Boolean value, you must make a choice.

**Review by callapa — CHANGES_REQUESTED** [review]

File: www/include/Administration/myAccount/DB-Func.php (hunk 1 above)
Comment (kduret) [reply www/include/Administration/myAccount/DB-Func.php]: ueless comment :)

**Review by callapa — APPROVED** [review]

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 9]
```diff
@@ -16452,3 +16452,9 @@ msgstr "Votre mot de passe expirera dans %s jours."
 
 msgid "Security policy not found. Please verify that your installation is valid"
 msgstr "Politique de sécurité non trouvée. Veuillez vérifier que votre installation est valide"
+
+msgid "Your autologin key should be different than your current password"
```
Comment (adr-mo) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:16456]: Simple comment here, but I guess the "should" implies that it is not possible to set an autologin key = password. Therefore "should" could be replaced by "must" as it is a requirement

File: www/include/Administration/myAccount/DB-Func.php [hunk 10]
```diff
@@ -219,3 +219,39 @@ function validatePasswordModification(array $fields)
 
     return count($errors) > 0 ? $errors : true;
 }
+
+/**
+ * @param array $fields
```
Comment (adr-mo) [inline www/include/Administration/myAccount/DB-Func.php:224]: Could it be more precise that array ?

File: www/include/configuration/configObject/contact/DB-Func.php [hunk 11]
```diff
@@ -1322,3 +1322,48 @@ function validatePasswordModification(array $fields)
 
     return count($errors) > 0 ? $errors : true;
 }
+
+/**
+ * Validate autologin key is not equal to a password
+ *
+ * @param array $fields
```
Comment (adr-mo) [inline www/include/configuration/configObject/contact/DB-Func.php:1329]: Could it be more precise that array ?

**Review by adr-mo — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**